### PR TITLE
chore: replace `ExecutionStage::read_block_with_senders` with `BlockProvider::block_with_senders`

### PR DESCRIPTION
**Pull request by joshieDo**
Moves `read_block_with_senders` from execution stage to `BlockProvider::block_with_senders`.

Also adds: 
`transactions_by_tx_range`
`senders_by_tx_range`

### CODE REVIEW

File: crates/storage/provider/src/providers/database/provider.rs [hunk 1]
```diff
@@ -1628,6 +1628,42 @@ impl<'this, TX: DbTx<'this>> BlockProvider for DatabaseProvider<'this, TX> {
     fn block_body_indices(&self, num: u64) -> Result<Option<StoredBlockBodyIndices>> {
         Ok(self.tx.get::<tables::BlockBodyIndices>(num)?)
     }
+
+    fn block_with_senders(&self, block_number: BlockNumber) -> Result<Option<BlockWithSenders>> {
```
Comment (onbjerg) [inline crates/storage/provider/src/providers/database/provider.rs:1632]: we need to note here that the tx's do not have valid hashes somehow

**Review by onbjerg — APPROVED** [review]
lgtm modulo note + lint and test failing

File: crates/storage/provider/src/providers/database/provider.rs (hunk 1 above)
Comment (joshieDo) [reply crates/storage/provider/src/providers/database/provider.rs:1638]: added some docs to the implementation on `DatabaseProvider`

**Review by gakonst — COMMENTED** [review]
Can we have 100% testing on each of these methods before merging? This is a great opportunity to generate proper regression tests for all these bindings, which previously bundled a lot of logic.

**Pull request merged**